staging/src/k8s.io/code-generator/cmd/validation-gen/validators/union.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validators
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"k8s.io/code-generator/cmd/validation-gen/annotations"
+	"k8s.io/gengo/v2"
+	"k8s.io/gengo/v2/generator"
+	"k8s.io/gengo/v2/types"
+)
+
+var discriminatedUnionValidator = types.Name{Package: libValidationPkg, Name: "DiscriminatedUnion"}
+var unionValidator = types.Name{Package: libValidationPkg, Name: "Union"}
+
+var newDiscriminatedUnionMembership = types.Name{Package: libValidationPkg, Name: "NewDiscriminatedUnionMembership"}
+var newUnionMembership = types.Name{Package: libValidationPkg, Name: "NewUnionMembership"}
+
+func init() {
+	AddToRegistry(InitUnionDeclarativeValidator)
+}
+
+func InitUnionDeclarativeValidator(c *generator.Context) DeclarativeValidator {
+	return &unionDeclarativeValidator{
+		universe: c.Universe,
+	}
+}
+
+type unionDeclarativeValidator struct {
+	universe types.Universe
+}
+
+const (
+	// +union and +unionDiscriminator tag are used by openapi-gen to publish x-kubernetes-union and x-kubernetes-discriminator
+	// extensions into Kubernetes published OpenAPI.
+	discriminatorTagName = "unionDiscriminator"
+	memberTagName        = "unionMember"
+)
+
+type discriminatorParams struct {
+	Union string `json:"union,omitempty"`
+}
+
+type memberParams struct {
+	Union      string `json:"union,omitempty"`
+	MemberName string `json:"memberName,omitempty"`
+}
+
+type union struct {
+	fields       []any
+	fieldMembers []any
+
+	discriminator       *string
+	discriminatorMember any
+}
+
+type unions map[string]*union
+
+func (us unions) getOrCreate(name string) *union {
+	var u *union
+	var ok bool
+	if u, ok = us[name]; !ok {
+		u = &union{}
+		us[name] = u
+	}
+	return u
+}
+
+func (c *unionDeclarativeValidator) ExtractValidations(t *types.Type, comments []string) (ValidatorGen, error) {
+	result := ValidatorGen{}
+
+	// TODO: handle multiple unions
+	// This really only works if we name each union
+	// so the most obvious approach is to track data by union
+	unions := unions{}
+
+	for _, member := range t.Members {
+		commentTags := gengo.ExtractCommentTags("+", member.CommentLines)
+		if tags, ok := commentTags[memberTagName]; ok {
+			if len(tags) != 1 {
+				return result, fmt.Errorf("must have one %q tag", memberTagName)
+			}
+			tag := tags[0]
+			var fieldName string
+			if jsonAnnotation, ok := annotations.LookupJSONTags(member); ok {
```

Comment:
> Shouldn't we be vocal about the error if we find a field which claims to be a unionMember but doesn't have JSON tags or doesn't have a name?

Added

> Edit: We need JSON tags pretty much everywhere, maybe we should just have some blanket assertion that validator code will never even get called if that isn't true? Since you (corrently) removed the panic - where can we do that?

This seems appropriate. OK if I put this in the TODO as a follow up?

Reply:
TODO SGTM